model/src/main/java/uk/gov/pay/commons/api/validation/ExternalJsonMetadataValidator.java

```diff
@@ -0,0 +1,72 @@
+package uk.gov.pay.commons.api.validation;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import java.util.Iterator;
+import java.util.Map;
+
+public class ExternalJsonMetadataValidator implements ConstraintValidator<ValidExternalJsonMetadata, JsonNode> {
+
+    private static final int MAXIMUM_KEYS = 10;
+    private static final int MAXIMUM_VALUE_LENGTH = 50;
+    private static final int MAXIMUM_KEY_LENGTH = 30;
+
+    @Override
+    public boolean isValid(JsonNode jsonNode, ConstraintValidatorContext constraintValidatorContext) {
+
+        if (jsonNode == null) {
+            return true;
+        }
+
+        constraintValidatorContext.disableDefaultConstraintViolation();
+
+        if (jsonNode.isArray()) {
+            addConstraint("must be an object of JSON key-value pairs", constraintValidatorContext);
+            return false;
+        }
+
+        int numberOfKeys = 0;
+        boolean isValid = true;
+        Iterator<Map.Entry<String, JsonNode>> keyValuePairs = jsonNode.fields();
+
+        while (keyValuePairs.hasNext()) {
+
+            Map.Entry<String, JsonNode> pair = keyValuePairs.next();
+
+            if (pair.getKey() == null) {
+                addConstraint(String.format("keys must not be null"), constraintValidatorContext);
+                isValid = false;
+            } else if (pair.getKey().length() < 1 || pair.getKey().length() > MAXIMUM_KEY_LENGTH) {
+                addConstraint(String.format("keys must be between 1 and %s characters long", MAXIMUM_KEY_LENGTH), constraintValidatorContext);
+                isValid = false;
+            }
+
+            if (++numberOfKeys > MAXIMUM_KEYS) {
+                addConstraint(String.format("cannot have more than %d key-value pairs", MAXIMUM_KEYS), constraintValidatorContext);
+                isValid = false;
+            }
```

Comment:
Will we get an error for each key-value pair over the maximum? should we only do this check once?

Reply:
Its a good point but we only get the error once. `validate` returns a `Set<ConstraintViolation<T>>` and having checked with trying to add 14 keys, it only returns one instance of that message since the `ConstraintViolation`s are considered equal. However maybe its not great to rely on that, so I could change the condition to...
```
if (++numberOfKeys  == MAXIMUM_KEYS + 1)
```
Which do you prefer? I'll update the test to attempt to add many more than the limit so its clear also.

Reply:
could we just move this check to after the iteration? might be clearer to read

Reply:
yes, that's a much better idea :)